File: NeuralNet/SnakeTile.m
//
//  SnakeTile.m
//  SnakeNet
//
//  Created by Maya Luna Celeste on 5/7/20.
//  Copyright © 2020 Maya Luna Celeste. All rights reserved.
//

#import "SnakeTile.h"

@implementation SnakeTile

-(instancetype)init{
    self.age = 0;
    self.tileType = TileTypeEmpty;
    return self;
}

- (id)copyWithZone:(NSZone *)zone{
    SnakeTile *copy = [SnakeTile new];
    copy.age = self.age;
    copy.tileType = self.tileType;
    copy.position = self.position;
    copy.next = [self.next copyWithZone:nil];
    copy.next.previous = self;
    return copy;
}

@end

File: NeuralNet/SnakeTile.swift
//
//  SnakeTile.swift
//  NeuralNet
//
//  Created by Maya Luna Celeste on 8/5/20.
//  Copyright © 2020 Maya Luna Celeste. All rights reserved.
//

import Cocoa

class SnakeTile: NSObject, NSCopying {
    enum TileType : Int {
        case empty = 0
        case head = 1
        case body = 2
        case food = 3
    }
    
    var tileType : TileType = TileType.empty
    var age : Int = 0
    var position : NSPoint = NSPoint()
    var next : SnakeTile!
    weak var previous : SnakeTile!
    
    func copy(with zone: NSZone? = nil) -> Any {
        var copy = SnakeTile()
        copy.age = age
        copy.tileType = tileType
        copy.position = position
        
        if next != nil {
            copy.next = next.copy() as? SnakeTile
            copy.next.previous = copy
        }
        
        return copy
    }
}
